src/equations/compressible_euler_3d.jl

```diff
@@ -1898,4 +1898,13 @@ end
 @inline function energy_internal(cons, equations::CompressibleEulerEquations3D)
     return energy_total(cons, equations) - energy_kinetic(cons, equations)
 end
+
+# State validation for Newton-bisection method of subcell IDP limiting
+@inline function Base.isvalid(u, equations::CompressibleEulerEquations3D)
+    p = pressure(u, equations)
+    if u[1] <= 0 || p <= 0
+        return false
+    end
+    return true
```

Comment:
Micro-optimization thought: What if you check density for positivity before you compute pressure? If density is negative, you can immediately `return false`